src/Renci.SshNet/SshCommand.cs

```diff
@@ -349,20 +352,49 @@ public string EndExecute(IAsyncResult asyncResult)
 
                 commandAsyncResult.EndCalled = true;
 
-                return Result;
+                if (!_isCancelled)
+                {
+                    return Result;
+                }
+
+                SetAsyncComplete();
+                throw new SshOperationCancelledException();
             }
         }
 
         /// <summary>
         /// Cancels command execution in asynchronous scenarios.
         /// </summary>
-        public void CancelAsync()
+        /// <param name="signalBeforeClose">should exit-signal be sent before attempting to close channel.</param>
+        /// <param name="forceKill">if true send SIGKILL instead of SIGTERM.</param>
+        /// <param name="timeout">how long to wait before stop waiting for command and close the channel.</param>
+        /// <returns>
+        /// Command Cancellation Task.
+        /// </returns>
+        /// <remarks>
+        /// <para>
+        /// After sending the exit-signal to the recipient, wait until either <paramref name="timeout"/> is exceeded
+        /// or the async result is signaled before signaling command cancellation.
+        /// If the exit-signal always results in the command being cancelled by the recipient, then <paramref name="timeout"/>
+        /// can be set to <see cref="Timeout.InfiniteTimeSpan"/> to wait until the async result is signaled.
+        /// </para>
+        /// </remarks>
+        public Task CancelAsync(bool signalBeforeClose = true, bool forceKill = false, TimeSpan timeout = default)
         {
-            if (_channel is not null && _channel.IsOpen && _asyncResult is not null)
+            if (signalBeforeClose)
             {
-                // TODO: check with Oleg if we shouldn't dispose the channel and uninitialize it ?
-                _channel.Dispose();
+                var signal = forceKill ? "KILL" : "TERM";
+                _ = _channel?.SendExitSignalRequest(signal, coreDumped: false, "Command execution has been cancelled.", "en");
             }
+
+            return Task.Run(() =>
+                         {
+                             var signaledElement = WaitHandle.WaitAny(new[] { _asyncResult.AsyncWaitHandle }, timeout);
+                             if (signaledElement == WaitHandle.WaitTimeout)
+                             {
+                                 _ = _commandCancelledWaitHandle?.Set();
+                             }
+                         });
```

Comment:
I think the version at 43a57e12 is better. Spawning a thread to block on a wait is not ideal.

I am thinking we should not try and wait for a response after all (I'm not convinced there is one now).

As for the callback, I am not sure but I think if we have cancelled the operation, we should not execute the callback. (if one cancels a Task-based operation with a CancellationToken, then we get a TaskCanceledException and continuations are not executed)